core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1244,6 +1245,74 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire streams fails.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+      Stream.eval(Deferred[F2, Unit]).flatMap { haltOuter =>
+        // consumes the upstream by sending a packet(stream and a ref) to the next queue and
+        // also does error handling/inturruption/switching by terminating the currently running
+        // inner stream
+        def runUpstream(next: Queue[F2, Option[(O, Deferred[F2, Unit])]],
+                        downstream: Queue[F2, Option[Either[Throwable, Chunk[O2]]]]) =
+          this.attempt
+            .interruptWhen(haltOuter.get.attempt)
+            .onFinalize(next.enqueue1(None))
+            .evalMap {
+              case Left(e) =>
+                haltRef.modify(x => None -> x.fold(F2.unit)(_.complete(()))).flatten *>
+                  downstream.enqueue1(Left(e).some)
+              case Right(x) =>
+                Deferred[F2, Unit].flatMap(halt =>
+                  haltRef.modify {
+                    case None       => halt.some -> F2.unit
+                    case Some(last) => halt.some -> last.complete(())
+                  }.flatten *> next.enqueue1((x -> halt).some))
+            }
+
+        // consumes the packets produced by runUpstream and drains the content to downstream
+        // upon error/interruption, the downstream will also be signalled for termination
+        def runInner(next: Queue[F2, Option[(O, Deferred[F2, Unit])]],
+                     downstream: Queue[F2, Option[Either[Throwable, Chunk[O2]]]]) =
+          next.dequeue.unNoneTerminate
+            .onFinalize(downstream.enqueue1(None))
+            .evalMap {
+              case (o, haltInner) =>
+                f(o).chunks
+                  .evalMap(x => downstream.enqueue1(Right(x).some))
+                  .interruptWhen(haltInner.get.attempt)
+                  .compile
+                  .drain
+                  .handleErrorWith(e => haltOuter.complete(()) *> downstream.enqueue1(Left(e).some))
```

Comment:
I'm wondering if this logic in `handleErrorWith` is necessary at all (not too sure, maybe it is).

If this stream fails, `runUpstream` will fail as well (it's in a `merge`), and `downstream` will fail as well (it's in `concurrently`). So is this related to the order of finalisers? I'm actually not super sure what's guaranteeing the order of finalisers here

Reply:
After replacing `evalMap` and lifting the error to a Stream I'm getting one test failure: `when inner stream fails, overall stream fails`. The reason was because no exception was thrown. Investigating now.

Reply:
Ah! I accidentally wrote `Stream( * : F2[Unit])`, should be `Stream.eval( * : F2[Unit])`. Damn, I had similar bugs like this when working with `merge` and `concurrently`.

Reply:
Ok, there was an actual bug in there where a `None` would get enqueued before an failure causing downstream to terminate too early. I guess that was why the semaphore was needed previously. I've rearranged that logic such that failures will always get enqueued first.